base/iobuffer.jl

```diff
@@ -317,118 +484,144 @@ function seek(io::GenericIOBuffer, n::Int)
+    # When append is true, new data is added after io.size, not io.ptr
+    existing_space = min(lastindex(io.data), io.maxsize) - (io.append ? io.size : io.ptr - 1)
+    if existing_space < nshort % Int
+        # Outline this function to make it more likely that ensureroom inlines itself
+        return ensureroom_slowpath(io, nshort)
     end
-    copyto!(io.data, 1, io.data, ptr, bytes_to_move)
-    io.size -= ptr - 1
-    io.ptr -= ptr - 1
+    return io
+end
+
+# Throw error (placed in this function to outline it) or reinit the buffer
+@noinline function ensureroom_reallocate(io::GenericIOBuffer, nshort::UInt)
+    io.writable || throw(ArgumentError("ensureroom failed, IOBuffer is not writeable"))
+    io.data = _similar_data(io, min(io.maxsize, nshort % Int))
+    io.reinit = false
     io.offset = 0
-    return
+    return io
 end
 
 @noinline function ensureroom_slowpath(io::GenericIOBuffer, nshort::UInt)
-    io.writable || throw(ArgumentError("ensureroom failed, IOBuffer is not writeable"))
-    if io.reinit
-        io.data = _similar_data(io, nshort % Int)
-        io.reinit = false
+    # Begin by zeroing out offset and check if that gives us room enough
+    if !iszero(io.offset)
+        nshort_after_zero_offset = (nshort % Int) - zero_offset!(io)
+        nshort_after_zero_offset < 1 && return io
+        nshort = nshort_after_zero_offset % UInt
     end
-    if !io.seekable
-        if !ismarked(io) && io.ptr > io.offset+1 && io.size <= io.ptr - 1
-            io.ptr = 1
-            io.size = 0
-            io.offset = 0
-        else
-            datastart = (ismarked(io) ? io.mark : io.ptr - io.offset)
-            if (io.size-io.offset+nshort > io.maxsize) ||
-                (datastart > 4096 && datastart > io.size - io.ptr) ||
-                (datastart > 262144)
-                # apply somewhat arbitrary heuristics to decide when to destroy
-                # old, read data to make more room for new data
-                compact(io)
-            end
+
+    data_len = min(lastindex(io.data), io.maxsize)
+
+    # Else, try to compact the data. To do this, the buffer must not be seekable.
+    # If it's seekable, the user can recover used data by seeking before ptr,
+    # and so we can't delete it.
+    if (!io.seekable && io.ptr > 1)
+        ptr = io.ptr
+        mark = io.mark
+        size = io.size
+        data = io.data
+        to_delete = (mark > -1 ? min(mark, ptr - 1) : ptr - 1)
+        # Only shift data if:
+        if (
+                # It will prevent us from having to resize buffer, or
+                to_delete >= nshort % Int ||
+                # We will recover at least 256 bytes, and at least 1/8th
+                # of the data buffer's total length
+                (to_delete > data_len >>> 3 && to_delete > 255)
+            )
+            copyto!(data, 1, data, to_delete + 1, size - to_delete)
+            io.ptr = ptr - to_delete
+            io.mark = max(-1, mark - to_delete)
```

Comment:
Reading the documentation for `mark`, this field value is not permitted to change here (we seem to have handled this wrong before too). I think you need to change offset instead, to essentially virtualize the position:
```suggestion
            io.offset -= to_delete
```

(although having a negative offset is slightly worrying, I think most computations actually should handle it fine already, but we should try to double-check that)

Reply:
I don't think a negative offset would be a correct solution to this problem. The offset already means something specific: The leading number of noncoding bytes in the data. Adding a separate meaning will lead to bugs.
However, perhaps the `offset` field can be stored in an encoded format as a `UInt`, where the top bit signals if the offset means "number of trailing unused bytes in the buffer", or it means "number of previously discarded, coding bytes after compacting". Only one of these will be needed at any one time, the offset will always be zero after compacting has happened.

Reply:
Maybe that's precisely what you meant - anyway, will give it a shot

Reply:
That seems close to what I meant, though I think the math usually works out such that those two different encodings definitions would always have the same numeric value: since we never access the data in the buffer before mark and that mark is relative to offset. This computationally is basically an OffsetArray, where accessing any data < mark is defined to throw an exception, so we don't actually need to keep that data, but may need to adjust the firstindex (aka offset) such that the computation of the mark location remains stable even if as it shifts the underlying indexes

Reply:
This has been addressed in (one of) the latest commits: https://github.com/JuliaLang/julia/pull/57570/commits/819d7506b51d385e534d043042567750687cf621. Might not be exactly what you envisioned, but it works and should be efficient.